internal/impl/snowflake/streaming/streaming.go

```diff
@@ -272,72 +271,132 @@ func (c *SnowflakeIngestionChannel) nextRequestID() string {
+	})
+	if err != nil {
+		return bdecPart{}, err
+	}
+	fileMetadata, err := readParquetMetadata(buf)
+	if err != nil {
+		return bdecPart{}, fmt.Errorf("unable to parse parquet metadata: %w", err)
+	}
+	done := time.Now()
+	return bdecPart{
+		unencryptedLen:  len(buf),
+		parquetFile:     buf,
+		parquetMetadata: fileMetadata,
+		stats:           combinedStats,
+		convertTime:     convertDone.Sub(convertStart),
+		serializeTime:   done.Sub(convertDone),
+	}, err
+}
+
 // InsertRows creates a parquet file using the schema from the data,
 // then writes that file into the Snowflake table
 func (c *SnowflakeIngestionChannel) InsertRows(ctx context.Context, batch service.MessageBatch) (InsertStats, error) {
-	stats := InsertStats{}
-	startTime := time.Now()
-	rows, err := constructRowGroup(batch, c.schema, c.transformers)
-	if err != nil {
-		return stats, err
+	insertStats := InsertStats{}
+	if len(batch) == 0 {
+		return insertStats, nil
 	}
-	// Prevent multiple channels from having the same bdec file (it must be unique)
-	// so add the ID of the channel in the upper 16 bits and then get 48 bits of
-	// randomness outside that.
+
+	startTime := time.Now()
+	// Prevent multiple channels from having the same bdec file (it must be globally unique)
+	// so add the ID of the channel in the upper 16 bits and then get 48 bits of randomness outside that.
 	fakeThreadID := (int(c.ID) << 48) | rand.N(1<<48)
 	blobPath := generateBlobPath(c.clientPrefix, fakeThreadID, int(c.requestIDCounter.Add(1)))
 	// This is extra metadata that is required for functionality in snowflake.
 	c.fileMetadata["primaryFileId"] = path.Base(blobPath)
-	c.buffer.Reset()
-	err = writeParquetFile(c.buffer, c.version, parquetFileData{
-		schema:   c.schema,
-		rows:     rows,
-		metadata: c.fileMetadata,
-	})
+	part, err := c.constructBdecPart(batch, c.fileMetadata)
 	if err != nil {
-		return stats, err
-	}
-	unencrypted := c.buffer.Bytes()
-	metadata, err := readParquetMetadata(unencrypted)
-	if err != nil {
-		return stats, fmt.Errorf("unable to parse parquet metadata: %w", err)
+		return insertStats, err
 	}
 	if debug {
-		_ = os.WriteFile("latest_test.parquet", unencrypted, 0o644)
+		_ = os.WriteFile("latest_test.parquet", part.parquetFile, 0o644)
 	}
-	unencryptedLen := len(unencrypted)
-	unencrypted = padBuffer(unencrypted, aes.BlockSize)
-	encrypted, err := encrypt(unencrypted, c.encryptionInfo.encryptionKey, blobPath, 0)
+
+	unencrypted := padBuffer(part.parquetFile, aes.BlockSize)
+	part.parquetFile, err = encrypt(unencrypted, c.encryptionInfo.encryptionKey, blobPath, 0)
 	if err != nil {
-		return stats, err
+		return insertStats, err
 	}
+
 	uploadStartTime := time.Now()
-	fileMD5Hash := md5.Sum(encrypted)
 	uploaderResult := c.uploader.Load()
 	if uploaderResult.err != nil {
-		return stats, fmt.Errorf("failed to acquire stage uploader: %w", uploaderResult.err)
+		return insertStats, fmt.Errorf("failed to acquire stage uploader: %w", uploaderResult.err)
```

Comment:
Nit: might be worth decorating the other returned errors as well.

Reply:
Done